Dynamic items in menu bar can cause crash

- VSCode Version: 1.26.0-ins
- OS Version: win7x64

```
Version: 1.26.0-insider
Commit: dae69dbf00357ff8a3fe740a1f211d66b0b4681c
Date: 2018-07-27T05:17:55.890Z
Electron: 2.0.5
Chrome: 61.0.3163.100
Node.js: 8.9.3
V8: 6.1.534.41
Architecture: x64
```

GIF with exts disabled:
![crash - menu bar](https://user-images.githubusercontent.com/24613274/43376386-34c8a0b2-93c3-11e8-8c81-3dedfd4236ca.gif)


1) force reload VS Code (or close and start)
2) after 1-2 seconds - open any menu bar when it's options are visible already
3) randomly move cursor above the options

**_Notice_**: with Disabled Extensions Mode it can be detected not so often, really rarely.
BUT in Mode with extensions enabled - every second-third trying to repro - makes a crash!

PS: btw where I can find logs of VS Code crashes?

@Tyriar 
```
PS C:\Users\stbatt\Source\vscode> code-insiders.cmd --verbose

[main 5:06:21 PM] Starting VS Code
[main 5:06:21 PM] from: c:\Users\stbatt\AppData\Local\Programs\Microsoft VS Code Insiders\resources\app
[main 5:06:21 PM] args: { _: [],
  help: false,
  h: false,
  version: false,
  v: false,
  wait: false,
  w: false,
  diff: false,
  d: false,
  add: false,
  a: false,
  goto: false,
  g: false,
  'new-window': false,
  n: false,
  'unity-launch': false,
  'reuse-window': false,
  r: false,
  'open-url': false,
  performance: false,
  p: false,
  'prof-startup': false,
  verbose: true,
  logExtensionHostCommunication: false,
  'disable-extensions': false,
  disableExtensions: false,
  'list-extensions': false,
  'show-versions': false,
  nolazy: false,
  issue: false,
  'skip-getting-started': false,
  'skip-release-notes': false,
  'sticky-quickopen': false,
  'disable-restore-windows': false,
  'disable-telemetry': false,
  'disable-updates': false,
  'disable-crash-reporter': false,
  'skip-add-to-recently-opened': false,
  status: false,
  s: false,
  'file-write': false,
  'file-chmod': false,
  'driver-verbose': false }
[main 5:06:21 PM] Resolving machine identifier...
[main 5:06:21 PM] Resolved machine identifier: 9db93396bd2759ea57c88403454e7401a6c5e52368525a73a6d77a802602bf66
[main 5:06:21 PM] update#setState idle
[main 5:06:21 PM] windowsManager#open
[main 5:06:21 PM] telemetry/optInStatus { optIn: true }
[main 5:06:21 PM] telemetry/update:win32SetupTarget { target: 'user' }
[18324:0730/170622.103:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[main 5:06:22 PM] windowsService#whenSharedProcessReady
[18324:0730/170622.129:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170622.137:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[main 5:06:22 PM] menubarService#updateMenubar 1
[main 5:06:22 PM] windowsService#getRecentlyOpened 1
[main 5:06:22 PM] menubarService#updateMenubar 1
[main 5:06:22 PM] windowsService#setRepresentedFilename 1
[18324:0730/170622.291:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170622.292:INFO:CONSOLE(1395)] "%cDEBUG", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[main 5:06:22 PM] windowsService#setRepresentedFilename 1
[18324:0730/170622.618:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[main 5:06:22 PM] windowsService#getWindowCount
[main 5:06:22 PM] IPC#vscode-workbenchLoaded
[main 5:06:22 PM] windowsService#getRecentlyOpened 1
[18324:0730/170622.862:INFO:CONSOLE(1395)] "%c INFO", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170623.054:INFO:CONSOLE(1395)] "%cDEBUG", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170623.080:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170623.083:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170623.083:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170623.083:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170623.084:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[main 5:06:23 PM] windowsService#startCrashReporter
[main 5:06:23 PM] menubarService#updateMenubar 1
[18324:0730/170623.355:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170623.355:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170623.355:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170623.355:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170623.356:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170623.491:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[main 5:06:23 PM] menubarService#updateMenubar 1
[main 5:06:23 PM] menubarService#updateMenubar 1
[18324:0730/170623.736:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[main 5:06:23 PM] windowsService#isFocused 1
[main 5:06:23 PM] menubarService#updateMenubar 1
[18324:0730/170625.058:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170625.058:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170625.059:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170625.059:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170625.059:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170625.059:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170625.059:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170625.060:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170625.061:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170625.061:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170625.061:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170625.062:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170625.164:INFO:CONSOLE(272)] "%c[Extension Host] %cActivating extension 'Equinusocio.vsc-material-theme' failed: Cannot find module './accents-setter'. %c(at Object.showMessage (c:\Users\stbatt\AppData\Local\Programs\Microsoft VS Code Insiders\resources\app\out\vs\workbench\node\extensionHostProcess.js:804:72))", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (272)
[18324:0730/170625.164:INFO:CONSOLE(272)] "%c[Extension Host] %cActivating extension `Equinusocio.vsc-material-theme` failed: ", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (272)
[18324:0730/170625.168:INFO:CONSOLE(272)] "%c[Extension Host] %cHere is the error stack: ", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (272)
[18324:0730/170625.169:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170625.169:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170625.213:INFO:CONSOLE(3220)] "Activating extension 'Equinusocio.vsc-material-theme' failed: Cannot find module './accents-setter'.", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (3220)
[18324:0730/170625.239:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170625.239:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170625.666:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170625.669:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[18324:0730/170625.689:INFO:CONSOLE(1395)] "%cTRACE", source: file:///C:/Users/stbatt/AppData/Local/Programs/Microsoft VS Code Insiders/resources/app/out/vs/workbench/workbench.main.js (1395)
[main 5:06:25 PM] menubarService#updateMenubar 1
```

--- a/src/vs/code/electron-main/app.ts
+++ b/src/vs/code/electron-main/app.ts
@@ -61,7 +61,6 @@ import { IMenubarService } from 'vs/platform/menubar/common/menubar';
 import { MenubarService } from 'vs/platform/menubar/electron-main/menubarService';
 import { MenubarChannel } from 'vs/platform/menubar/node/menubarIpc';
 import { ILabelService, RegisterFormatterEvent } from 'vs/platform/label/common/label';
-import { CodeMenu } from 'vs/code/electron-main/menus';
 import { hasArgs } from 'vs/platform/environment/node/argv';
 import { RunOnceScheduler } from 'vs/base/common/async';
 import { registerContextMenuListener } from 'vs/base/parts/contextmenu/electron-main/contextmenu';
@@ -592,23 +591,6 @@ export class CodeApplication {
 					});
 				}
 			}
-		}
-
-		// Install Menu
-		const instantiationService = accessor.get(IInstantiationService);
-		const configurationService = accessor.get(IConfigurationService);
-
-		// Initial value sets the value for macOS. false indicates using Dynamic Menus
-		// Eventually this will not be necessary as the dynamic menu will takeover
-		let createNativeMenu = false;
-		if (platform.isLinux) {
-			createNativeMenu = configurationService.getValue<string>('window.titleBarStyle') !== 'custom';
-		} else if (platform.isWindows) {
-			createNativeMenu = configurationService.getValue<string>('window.titleBarStyle') === 'native';
-		}
-
-		if (createNativeMenu) {
-			instantiationService.createInstance(CodeMenu);
 		}
 
 		// Jump List

--- a/src/vs/platform/menubar/electron-main/menubar.ts
+++ b/src/vs/platform/menubar/electron-main/menubar.ts
@@ -211,7 +211,8 @@ export class Menubar {
 	}
 
 	private install(): void {
-
+		// Store old menu in our array to avoid GC to collect the menu and crash. See #55347
+		// TODO@sbatten Remove this when fixed upstream by Electron
 		const oldMenu = Menu.getApplicationMenu();
 		if (oldMenu) {
 			this.oldMenus.push(oldMenu);

--- a/src/vs/platform/menubar/electron-main/menubar.ts
+++ b/src/vs/platform/menubar/electron-main/menubar.ts
@@ -38,6 +38,8 @@ export class Menubar {
 	private menuUpdater: RunOnceScheduler;
 	private menuGC: RunOnceScheduler;
 
+	// Array to keep menus around so that GC doesn't cause crash as explained in #55347
+	// TODO@sbatten Remove this when fixed upstream by Electron
 	private oldMenus: Menu[];
 
 	private menubarMenus: IMenubarData;
@@ -59,7 +61,7 @@ export class Menubar {
 	) {
 		this.menuUpdater = new RunOnceScheduler(() => this.doUpdateMenu(), 0);
 
-		this.menuGC = new RunOnceScheduler(() => { this.oldMenus = this.oldMenus.slice(this.oldMenus.length - 1); }, 10000);
+		this.menuGC = new RunOnceScheduler(() => { this.oldMenus = []; }, 10000);
 
 		this.menubarMenus = this.stateService.getItem<IMenubarData>(Menubar.lastKnownMenubarStorageKey) || Object.create(null);
 

--- a/src/vs/platform/menubar/electron-main/menubar.ts
+++ b/src/vs/platform/menubar/electron-main/menubar.ts
@@ -36,6 +36,9 @@ export class Menubar {
 	private closedLastWindow: boolean;
 
 	private menuUpdater: RunOnceScheduler;
+	private menuGC: RunOnceScheduler;
+
+	private oldMenus: Menu[];
 
 	private menubarMenus: IMenubarData;
 
@@ -56,6 +59,8 @@ export class Menubar {
 	) {
 		this.menuUpdater = new RunOnceScheduler(() => this.doUpdateMenu(), 0);
 
+		this.menuGC = new RunOnceScheduler(() => { this.oldMenus = this.oldMenus.slice(this.oldMenus.length - 1); }, 10000);
+
 		this.menubarMenus = this.stateService.getItem<IMenubarData>(Menubar.lastKnownMenubarStorageKey) || Object.create(null);
 
 		this.keybindings = this.stateService.getItem<IMenubarData>(Menubar.lastKnownAdditionalKeybindings) || Object.create(null);
@@ -63,6 +68,8 @@ export class Menubar {
 		this.addFallbackHandlers();
 
 		this.closedLastWindow = false;
+
+		this.oldMenus = [];
 
 		this.install();
 
@@ -203,6 +210,11 @@ export class Menubar {
 
 	private install(): void {
 
+		const oldMenu = Menu.getApplicationMenu();
+		if (oldMenu) {
+			this.oldMenus.push(oldMenu);
+		}
+
 		// If we don't have a menu yet, set it to null to avoid the electron menu.
 		// This should only happen on the first launch ever
 		if (Object.keys(this.menubarMenus).length === 0) {
@@ -293,15 +305,6 @@ export class Menubar {
 			menubar.append(macWindowMenuItem);
 		}
 
-		// Preferences
-		if (!isMacintosh) {
-			const preferencesMenu = new Menu();
-			const preferencesMenuItem = new MenuItem({ label: this.mnemonicLabel(nls.localize({ key: 'mPreferences', comment: ['&& denotes a mnemonic'] }, "&&Preferences")), submenu: preferencesMenu });
-
-			this.setMenuById(preferencesMenu, 'Preferences');
-			menubar.append(preferencesMenuItem);
-		}
-
 		// Help
 		const helpMenu = new Menu();
 		const helpMenuItem = new MenuItem({ label: this.mnemonicLabel(nls.localize({ key: 'mHelp', comment: ['&& denotes a mnemonic'] }, "&&Help")), submenu: helpMenu, role: 'help' });
@@ -314,6 +317,9 @@ export class Menubar {
 		} else {
 			Menu.setApplicationMenu(null);
 		}
+
+		// Dispose of older menus after some time
+		this.menuGC.schedule();
 	}
 
 	private setMacApplicationMenu(macApplicationMenu: Electron.Menu): void {

--- a/src/vs/platform/menubar/electron-main/menubarService.ts
+++ b/src/vs/platform/menubar/electron-main/menubarService.ts
@@ -8,7 +8,6 @@ import { Menubar } from 'vs/platform/menubar/electron-main/menubar';
 import { ILogService } from 'vs/platform/log/common/log';
 import { TPromise } from 'vs/base/common/winjs.base';
 import { IInstantiationService } from 'vs/platform/instantiation/common/instantiation';
-import { isMacintosh } from 'vs/base/common/platform';
 
 export class MenubarService implements IMenubarService {
 	_serviceBrand: any;
@@ -20,10 +19,7 @@ export class MenubarService implements IMenubarService {
 		@ILogService private logService: ILogService
 	) {
 		// Install Menu
-		// TODO@sbatten Only do this for macOS as the dynamic menus are currently stable. #55347
-		if (isMacintosh) {
-			this._menubar = this.instantiationService.createInstance(Menubar);
-		}
+		this._menubar = this.instantiationService.createInstance(Menubar);
 	}
 
 	updateMenubar(windowId: number, menus: IMenubarData, additionalKeybindings?: Array<IMenubarKeybinding>): TPromise<void> {

--- a/src/vs/workbench/browser/parts/titlebar/menubarControl.ts
+++ b/src/vs/workbench/browser/parts/titlebar/menubarControl.ts
@@ -468,8 +468,7 @@ export class MenubarControl extends Disposable {
 	private doSetupMenubar(): void {
 		if (!isMacintosh && this.currentTitlebarStyleSetting === 'custom') {
 			this.setupCustomMenubar();
-		} else if (isMacintosh) {
-			// TODO@sbatten currently limiting dynamic native menus to macOS #55347
+		} else {
 			// Send menus to main process to be rendered by Electron
 			const menubarData = {};
 			if (this.getMenubarMenus(menubarData)) {

